File: src/platform/music.ts
import { call, on } from './invoke';
import type { BridgeStatus, LyricsData, MusicAction, MusicState } from './types';

export const musicApi = {
  poll: () => call<MusicState>('music_poll'),

  /** 轮询外的即时推送 */
  onState: (cb: (state: MusicState) => void) => on<MusicState>('music:state', cb),

  control: (action: MusicAction, level?: number) =>
    call<string>('music_control', { action, level: level ?? null }),

  seek: (positionMs: number) => call<boolean>('music_seek', { positionMs }),

  /** id 不匹配返回 null */
  lyrics: (id: string) => call<LyricsData | null>('music_lyrics', { id }),

  bridgeStatus: () => call<BridgeStatus>('music_bridge_status'),
};

File: src-tauri/src/ipc/music.rs
use island_core::{LyricsData, MusicAction, MusicState};

use super::off_thread;
use crate::error::AppResult;
use crate::services;
use crate::services::music::BridgeStatus;

#[tauri::command]
pub async fn music_state() -> AppResult<MusicState> {
    off_thread(|| Ok(services::music::current_state())).await
}

#[tauri::command]
pub async fn music_control(action: MusicAction, level: Option<i64>) -> AppResult<String> {
    off_thread(move || services::music::control(action, level)).await
}

#[tauri::command]
pub async fn music_seek(position_ms: i64) -> AppResult<bool> {
    off_thread(move || Ok(services::music::seek(position_ms))).await
}

/// id 已切歌时回 None
#[tauri::command]
pub async fn music_lyrics(id: String) -> AppResult<Option<LyricsData>> {
    off_thread(move || Ok(services::music::lyrics(&id))).await
}

#[tauri::command]
pub async fn music_bridge_status() -> AppResult<BridgeStatus> {
    off_thread(|| Ok(services::music::bridge_status())).await
}
